File: src/dexter/frontend/clang/src/ir/bool/BinaryExpr.cpp
/*
 * BinaryExpr.cpp
 *
 *  Created on: Mar 24, 2018
 *      Author: maazsaf
 */

#include "ir/bool/BinaryExpr.h"

Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::AND = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::OR = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::EQ = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::NEQ = new Op();

Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::LT = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::LE = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::GE = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::GT = new Op();

Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::PLUS = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::MINUS = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::MULT = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::DIV = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::MOD = new Op();

Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::SHL = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::LSHR = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::ASHR = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::BAND = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::BOR = new Op();
Dexter::BinaryExpr::Op * Dexter::BinaryExpr::Op::BXOR = new Op();

Dexter::BinaryExpr::BinaryExpr (Dexter::Expr * left, Op * op, Dexter::Expr * right) : Dexter::Expr(NULL)
{
  JNIEnv * e = Util::env;
  jclass c = e->FindClass("dexter/ir/bool/BinaryExpr");
  jmethodID m = e->GetMethodID(c, "<init>", "(Ldexter/ir/Expr;Ldexter/ir/bool/BinaryExpr$Op;Ldexter/ir/Expr;)V");
  this->mObj = e->NewObject(c, m, left->obj(), op->obj(), right->obj());
}

void Dexter::BinaryExpr::init ()
{
  JNIEnv * e = Util::env;
  jclass ts = e->FindClass("dexter/ir/bool/BinaryExpr$Op");

  const char * opClass = "Ldexter/ir/bool/BinaryExpr$Op;";

  Op::AND->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "AND", opClass));
  Op::OR->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "OR", opClass));
  Op::EQ->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "EQ", opClass));
  Op::NEQ->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "NEQ", opClass));

  Op::LT->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "LT", opClass));
  Op::LE->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "LE", opClass));
  Op::GE->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "GE", opClass));
  Op::GT->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "GT", opClass));

  Op::PLUS->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "PLUS", opClass));
  Op::MINUS->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "MINUS", opClass));
  Op::MULT->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "MULT", opClass));
  Op::DIV->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "DIV", opClass));
  Op::MOD->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "MOD", opClass));

  Op::SHL->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "SHL", opClass));
  Op::LSHR->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "LSHR", opClass));
  Op::ASHR->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "ASHR", opClass));
  Op::BAND->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "BAND", opClass));
  Op::BOR->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "BOR", opClass));
  Op::BXOR->mObj = e->GetStaticObjectField(ts, e->GetStaticFieldID(ts, "BXOR", opClass));
}

File: src/dexter/ir/bool/BinaryExpr.java
package dexter.ir.bool;

import dexter.ir.Expr;
import dexter.ir.Visitor;
import dexter.ir.type.TypesFactory;

/**
 * Created by Maaz Ahmad on 6/25/19.
 */
public class BinaryExpr extends Expr
{
  public static final class Op
  {
    private String symbol;

    private Op(String symbol) { this.symbol = symbol; }

    public static final Op AND = new Op("&&");
    public static final Op OR = new Op("||");
    public static final Op EQ = new Op("=");
    public static final Op NEQ = new Op("!=");

    public static final Op LT = new Op("<");
    public static final Op LE = new Op("<=");
    public static final Op GE = new Op(">=");
    public static final Op GT = new Op(">");

    public static final Op PLUS = new Op("+");
    public static final Op MINUS = new Op("-");
    public static final Op MULT = new Op("*");
    public static final Op DIV = new Op("/");
    public static final Op MOD = new Op("%");

    public static final Op SHL = new Op("<<");
    public static final Op LSHR = new Op(">>");
    public static final Op ASHR = new Op(">>>");
    public static final Op BAND = new Op("&");
    public static final Op BOR = new Op("|");
    public static final Op BXOR = new Op("^");

    public static final Op IMPLY = new Op("-->");

    @Override public String toString () { return symbol; }
  }

  protected Expr left;
  protected Op op;
  protected Expr right;

  public BinaryExpr(Expr left, Op op, Expr right)
  {
    this.left = left;
    this.op = op;
    this.right = right;

    if (op == BinaryExpr.Op.AND) { this.type = TypesFactory.Bool; }
    else if (op == BinaryExpr.Op.OR) { this.type = TypesFactory.Bool; }
    else if (op == BinaryExpr.Op.EQ) { this.type = TypesFactory.Bool; }
    else if (op == BinaryExpr.Op.NEQ) { this.type = TypesFactory.Bool; }

    else if (op == BinaryExpr.Op.LT) { this.type = TypesFactory.Bool; }
    else if (op == BinaryExpr.Op.LE) { this.type = TypesFactory.Bool; }
    else if (op == BinaryExpr.Op.GE) { this.type = TypesFactory.Bool; }
    else if (op == BinaryExpr.Op.GT) { this.type = TypesFactory.Bool; }

    else if (op == BinaryExpr.Op.IMPLY) { this.type = TypesFactory.Bool; }

    // left and right should have the same type now
    else if (op == BinaryExpr.Op.PLUS) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.MINUS) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.MULT) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.DIV) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.MOD) { this.type = left().type(); }

    else if (op == BinaryExpr.Op.SHL) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.LSHR) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.ASHR) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.BAND) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.BOR) { this.type = left().type(); }
    else if (op == BinaryExpr.Op.BXOR) { this.type = left().type(); }
  }

  public Expr left () { return left; }
  public Op op () { return op; }
  public Expr right () { return right; }


  @Override
  public <T> T accept(Visitor<T> p)
  {
    return p.visit(this);
  }

  @Override
  public boolean equals (Object o)
  {
    if (o instanceof BinaryExpr)
      return ((BinaryExpr)o).left.equals(left) &&
              ((BinaryExpr)o).op.equals(op) &&
              ((BinaryExpr)o).right.equals(right);
    else
      return false;
  }
}